cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringAgentService.java

```diff
@@ -91,11 +102,17 @@
   private final TransactionRunner transactionRunner;
   private final Map<String, String> lastMessageIds;
   private final ProgramStateWriter programStateWriter;
+  private final MessageFetcher messageFetcher;
+  private final ProgramRunRecordFetcher runRecordFetcher;
   private final RemoteAuthenticator remoteAuthenticator;
+  private final String programUpdateTopic;
+  private final int programUpdateFetchSize;
+  private String lastProgramUpdateMessageId;
 
   @Inject
   TetheringAgentService(CConfiguration cConf, TransactionRunner transactionRunner, TetheringStore store,
-                        ProgramStateWriter programStateWriter,
+                        ProgramStateWriter programStateWriter, MessagingService messagingService,
```

Comment:
There is a AbstractNotificationSubscriberService that has a bunch of base code for reading messages from a topic and managing subscriber state. Can you look into whether it makes sense for this class to extend that one?